File: applications/examples/python/Truss1D2N.py
import nuto
import numpy as np


class Material:
    youngsModulus = 20000.


class Geometry:
    lx = 70.
    ly = 3.1415
    lz = 0.42
    numElements = 10


class BoundaryCondition:
    force = 12.
    displacement = 0.1


def SetupGeometry(structure):
    # create a 1D Truss mesh, add a linear displacement interpolation and update the mesh
    meshInfo = nuto.MeshGenerator.Grid(structure, [Geometry.lx], [Geometry.numElements])
    structure.InterpolationTypeAdd(meshInfo[1], "displacements", "equidistant1")
    structure.ElementTotalConvertToInterpolationType()

    # create and set section
    sectionId = structure.SectionCreate("TRUSS")
    structure.SectionSetArea(sectionId, Geometry.ly * Geometry.lz)
    structure.ElementTotalSetSection(sectionId)


def SetupMaterial(structure):
    # create and set material law
    materialId = structure.ConstitutiveLawCreate("Linear_Elastic_Engineering_Stress")
    structure.ConstitutiveLawSetParameterDouble(materialId, "Youngs_Modulus", Material.youngsModulus)
    structure.ElementTotalSetConstitutiveLaw(materialId)


def SetupBoundaryConditions(structure, BCType):
    nodeSelectionTolerance = 1.e-6
    xDirection = np.array([1.0])

    # fix left node at x = [0]
    nodeLeft = structure.NodeGetIdAtCoordinate(np.array([0.]), nodeSelectionTolerance)
    prescribedDisplacement = 0.
    structure.ConstraintLinearSetDisplacementNode(nodeLeft, xDirection, prescribedDisplacement)

    # apply nonzero BC at x = length
    nodeRight = structure.NodeGetIdAtCoordinate(np.array([Geometry.lx]), nodeSelectionTolerance)
    if BCType == "DisplacmentBC":
        structure.ConstraintLinearSetDisplacementNode(nodeRight, xDirection, BoundaryCondition.displacement)
    if BCType == "ForceBC":
        structure.SetNumLoadCases(1)
        structure.LoadCreateNodeForce(0, nodeRight, xDirection, BoundaryCondition.force)


def Solve(structure):
    structure.SolveGlobalSystemStaticElastic(0)
    intGradient = structure.BuildGlobalInternalGradient()
    extGradient = structure.BuildGlobalExternalLoadVector(0)
    residual = intGradient.J.Get("Displacements") - extGradient.J.Get("Displacements")
    print("residual: {0}".format(np.linalg.norm(residual)))
    print("(should be _very_ close to zero.)")


def Visualize(structure, file):
    visualizationGroup = structure.GroupGetElementsTotal()

    structure.AddVisualizationComponent(visualizationGroup, "Displacements")
    structure.AddVisualizationComponent(visualizationGroup, "EngineeringStrain")
    structure.AddVisualizationComponent(visualizationGroup, "EngineeringStress")
    structure.ExportVtkDataFileElements(file)


def Run(BCType):
    # create one-dimensional structure
    structure = nuto.Structure(1)

    SetupGeometry(structure)
    SetupMaterial(structure)
    SetupBoundaryConditions(structure, BCType)

    Solve(structure)

    Visualize(structure, "Truss1D2N_" + BCType + ".vtk")


Run("DisplacmentBC")
Run("ForceBC")

File: applications/examples/c++/Truss1D2N.cpp
#include <iostream>

#include "math/SparseMatrixCSRGeneral.h"
#include "math/SparseDirectSolverMUMPS.h"

#include "mechanics/MechanicsEnums.h"
#include "mechanics/constitutive/laws/AdditiveInputExplicit.h"
#include "mechanics/constraints/ConstraintCompanion.h"
#include "mechanics/dofSubMatrixStorage/BlockScalar.h"
#include "mechanics/sections/SectionTruss.h"
#include "mechanics/structures/unstructured/Structure.h"
#include "mechanics/structures/StructureOutputBlockVector.h"
#include "visualize/VisualizeEnum.h"

using namespace NuTo;

int main()
{
    // definitions
    double YoungsModulus = 20000.;
    double Area = 100. * 100.;
    double Length = 1000.;
    int NumElements = 10;
    double Force = 1.;
    bool EnableDisplacementControl = false;
    double BoundaryDisplacement = 0.1;

    // create one-dimensional structure
    Structure structure(1);

    // create section
    auto Section1 = SectionTruss::Create(Area);

    // create material law
    auto Material1 = structure.ConstitutiveLawCreate("Linear_Elastic_Engineering_Stress");
    structure.ConstitutiveLawSetParameterDouble(Material1, Constitutive::eConstitutiveParameter::YOUNGS_MODULUS,
                                                YoungsModulus);

    // create nodes
    Eigen::VectorXd nodeCoordinates(1);
    for (int node = 0; node < NumElements + 1; node++)
    {
        std::cout << "create node: " << node << " coordinates: " << node * Length / NumElements << std::endl;
        nodeCoordinates(0) = node * Length / NumElements;
        structure.NodeCreate(node, nodeCoordinates);
    }

    int InterpolationType = structure.InterpolationTypeCreate("Truss1D");
    structure.InterpolationTypeAdd(InterpolationType, Node::eDof::COORDINATES, Interpolation::eTypeOrder::EQUIDISTANT1);
    structure.InterpolationTypeAdd(InterpolationType, Node::eDof::DISPLACEMENTS,
                                   Interpolation::eTypeOrder::EQUIDISTANT1);

    // create elements
    std::vector<int> elementIncidence(2);
    for (int element = 0; element < NumElements; element++)
    {
        std::cout << "create element: " << element << " nodes: " << element << "," << element + 1 << std::endl;
        elementIncidence[0] = element;
        elementIncidence[1] = element + 1;
        structure.ElementCreate(InterpolationType, elementIncidence);
        structure.ElementSetSection(element, Section1);
        structure.ElementSetConstitutiveLaw(element, Material1);
    }

    structure.ElementTotalConvertToInterpolationType();

    // set boundary conditions and loads
    auto& origin = structure.NodeGetAtCoordinate(Eigen::Matrix<double, 1, 1>::Zero());
    structure.Constraints().Add(Node::eDof::DISPLACEMENTS, Constraint::Value(origin));
    if (EnableDisplacementControl)
    {
        std::cout << "Displacement control" << std::endl;
        auto& lastNode = *structure.NodeGetNodePtr(NumElements);
        structure.Constraints().Add(Node::eDof::DISPLACEMENTS, Constraint::Value(lastNode, BoundaryDisplacement));
    }
    else
    {
        std::cout << "Load control" << std::endl;
        Eigen::VectorXd direction(1);
        direction(0) = 1;
        structure.LoadCreateNodeForce(NumElements, direction, Force);
    }

    // start analysis
    structure.SolveGlobalSystemStaticElastic();
    auto residual = structure.BuildGlobalInternalGradient() - structure.BuildGlobalExternalLoadVector();

    std::cout << "residual: " << residual.J.CalculateNormL2() << std::endl;

    // visualize results
    int visualizationGroup = structure.GroupCreate(eGroupId::Elements);
    structure.GroupAddElementsTotal(visualizationGroup);

    structure.AddVisualizationComponent(visualizationGroup, eVisualizeWhat::DISPLACEMENTS);
    structure.AddVisualizationComponent(visualizationGroup, eVisualizeWhat::ENGINEERING_STRAIN);
    structure.AddVisualizationComponent(visualizationGroup, eVisualizeWhat::ENGINEERING_STRESS);
    structure.ExportVtkDataFileElements("Truss1D2N.vtk");

    return 0;
}
